[Site Admin]: Add NavigableRegion component

retrofox opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Introduces a new `<NavigableRegion />` component that creates accessible regions with keyboard navigation support.
This component follows Gutenberg's pattern for creating navigable sections in the admin interface, providing proper ARIA attributes and keyboard interaction capabilities.

The component is flexible, allowing different HTML elements via the `as` prop while maintaining consistent accessibility features across the interface.

Related to ARC-27

## Proposed Changes

* [Site Admin]: Add NavigableRegion component

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

*

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Visual review will be enough for now. In the ARC-25 follow-up, we will put this component into practice.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

retrofox committed:
Introduce first approach of NavigableRegion cmp
Create new component for accessible keyboard navigation regions with ARIA support

retrofox committed:
improve NavigableRegion component types

retrofox committed:
add JSDoc to NavigableRegion component

retrofox committed:
changelog

retrofox committed:
reorganize interface cmps - expose NavigableRegion cpm

tyxla commented:
Ah, bummer we're duplicating yet another existing component. I wish we could just use the one from `@wordpress/interface`.

tyxla commented:
Yep, unfortunately. I have a special opinion about components that we need to use externally but we decided to keep private.  

I'll defer to @mirka / @ciampo to make the call here.

retrofox commented:
<!-- linear:isThreadRoot -->

The plan is to get everything ready so we can switch to the core one as soon as it becomes public

ciampo commented:
I don't see an alternative, given that the component is not exported from `@wordpress/interface`, unfortunately.

Although I'm slightly confused by this sentence:

> The plan is to get everything ready so we can switch to the core one as soon as it becomes public

I was under the impression that the goal for the `@automattic/site-admin` package was "visual consistency" with Gutenberg, and that maintaining compatibility with the APIs of the original Gutenberg package (and tweaking code as little as possible) was NOT a goal, like Riad clarified [in this comment](https://github.com/Automattic/wp-calypso/pull/100493#issuecomment-2692605116).

If that's the case, switching to core as soon as it becomes public may not be possible as our version may have diverged signifcantly.

retrofox commented:
> I don't see an alternative, given that the component is not exported from `@wordpress/interface`, unfortunately.
> 
> Although I'm slightly confused by this sentence:
> 
> > The plan is to get everything ready so we can switch to the core one as soon as it becomes public
> 
> I was under the impression that the goal for the `@automattic/site-admin` package was "visual consistency" with Gutenberg, and that maintaining compatibility with the APIs of the original Gutenberg package (and tweaking code as little as possible) was NOT a goal, like Riad clarified [in this comment](https://github.com/Automattic/wp-calypso/pull/100493#issuecomment-2692605116).
> 
> If that's the case, switching to core as soon as it becomes public may not be possible as our version may have diverged signifcantly.


Yes, that makes sense. But if, for any reason, core exposes resources that are copies of them — like the router package, for instance — we can switch to those instead.

ciampo commented on packages/site-admin/src/interface/components/navigable-region/index.tsx:
```diff
@@ -0,0 +1,41 @@
+		return (
+			<Tag
+				ref={ ref }
+				className={ clsx( 'interface-navigable-region', className ) }
```
Let's change the classname, using the same CSS prefix used in the rest of the package.

ciampo commented on packages/site-admin/src/interface/components/navigable-region/index.tsx:
```diff
@@ -0,0 +1,41 @@
+);
+
+NavigableRegion.displayName = 'NavigableRegion';
+export default NavigableRegion;
```
This is a pet peeve of mine, but — is there any reason why we're using default exports? I personally prefer named exports since they make code searching a lot more predictable.

retrofox replied:
simply following the core implementation. Let me update it

retrofox committed:
remove component className
This className is not used for now. We can add it in follow-ups if it's required.

retrofox committed:
Update NavigableRegion export to named export

retrofox committed:
fix changelog tracks

ciampo commented on packages/site-admin/src/interface/components/navigable-region/index.tsx:
```diff
@@ -0,0 +1,31 @@
+			</Tag>
+		);
+	}
+);
```
There is a bug and a few improvements that can be made:
- The `className` prop is currently extracted from the `props` but never assigned to the `Tag` component;
- There's no need to extract the `ariaLabel` and `children` props to just pass them to `Tag`;
- Since we're accepting all `React.HTMLAttributes< HTMLElement`, there's no need to explicitly add `className` to the props;
- As a nice to have, we can re-use the aria-label prop type from `React.HTMLAttributes< HTMLElement`
- As another nice to have, we can add the default values for `role` and `tabIndex` directly in the prop restructuring

We can fix/improve the code by rewriting it this way:

```suggestion
type NavigableRegionProps = {
	children: React.ReactNode;
	ariaLabel: NonNullable< HTMLElement[ 'ariaLabel' ] >;
	as?: React.ElementType;
} & React.HTMLAttributes< HTMLElement >;

export const NavigableRegion = forwardRef< HTMLElement, NavigableRegionProps >(
	( { as: Tag = 'div', role = 'region', tabIndex = -1, ...props }, ref ) => {
		return <Tag ref={ ref } role={ role } tabIndex={ tabIndex } { ...props } />;
	}
);
```

ciampo reviewed (approved):
Pre-approving, given that I won't be very responsive for the rest of the week.

Feel free to merge after addressing https://github.com/Automattic/wp-calypso/pull/102361#discussion_r2033140046

retrofox committed:
Update packages/site-admin/src/interface/components/navigable-region/index.tsx

Co-authored-by: Marco Ciampini <marco.ciampini@automattic.com>
